### PR TITLE
Delfin NAS resource model analysis for Filesystem, Qtree, Share and Quota

### PR DESCRIPTION
**Pull request by joseph-v**
**What type of PR is this?**

Delfin NAS resource model FS, Qtree, Quota analysis

> /kind design

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:

**Special notes for your reviewer**:

### CODE REVIEW

File: Analysis/DataStorageProjects/Delfin/NASModels/NASResources.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+## NAS Resource Model Analysis
+
+Following are the resource models derived from analyzing some the leading storage vendors including DELL, Hitachi, HP, Huawei, IBM etc.
+
+### NAS Resource Model implementation diagram
+
+![NAS Resource Model Diagram](NASResourceModel.jpg)
+
+### Filesystem
```
Comment (kumarashit) [inline Analysis/DataStorageProjects/Delfin/NASModels/NASResources.md:37]: What is the relation from NAS Volume to Filesystem?

File: Analysis/DataStorageProjects/Delfin/NASModels/NASResources.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+-- | -- | --
+id | string | UUID of the Qtree
+name | string | Name of the Qtree
+storage_id | string | Delfin id of the associated storage
+native_qtree_id | string | Original Qtree id in the device
+filesystem_id | string | Delfin id of the parent filesystem
+quota_id | string | Delfin id of the quota applied to Qtree
+path | string | Path of the Qtree
+used_capacity | long | Capacity used
+nfs_exports | long | Number of NFS exports
```
Comment (kumarashit) [inline Analysis/DataStorageProjects/Delfin/NASModels/NASResources.md:68]: How is it different from Filesystem attributes?

File: Analysis/DataStorageProjects/Delfin/NASModels/NASResources.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+free_capacity | long | Free capacity in bytes of the Filesystem
+status | enum | Running status of the Filesystem (normal, offline, unknown)
+type | string | Filesystem type (thick, thin)
+deduplication | bool | Filesystem deduplication (true, false)
+compression | bool | Filesystem compression (true, false)
+security_mode | string | Security Mode of Filesystem (mixed, native, windows, unix)
+nfs_exports | long | Number of NFS exports
+cifs_exports | long | Number of CIFS exports
+
+### Qtree
```
Comment (kumarashit) [inline Analysis/DataStorageProjects/Delfin/NASModels/NASResources.md:56]: Qtree can be exposed as iSCSI LUN as well as Share. How is that managed?

**Comment by skdwriting** [comment]
1. Should we call this as NAS or file storage support ? (NAS is Network Attached Storage which is a file type storage)
2. Can we have the analysis inputs in the document for "Following are the resource models derived from analyzing some the leading storage vendors including DELL, Hitachi, HP, Huawei, IBM etc." ?
3. "security_mode" of Qtree, do we have any impact on the monitoring access security?!
4. The filesystem, Qtree, Quota -- all these data models are unified from different vendors?! (There are exclusive parameters as well?)
5. From share side, what are we targetting?
6. Do we have any recommendations for the APIs when we considering a new type of (file) storage?! -- OR we are fitting these models to existing APIs?! -- This is a critical point and can impact the overall arch

**Comment by joseph-v** [comment]
> 1. Should we call this as NAS or file storage support ? (NAS is Network Attached Storage which is a file type storage)
> 2. Can we have the analysis inputs in the document for "Following are the resource models derived from analyzing some the leading storage vendors including DELL, Hitachi, HP, Huawei, IBM etc." ?
> 3. "security_mode" of Qtree, do we have any impact on the monitoring access security?!
> 4. The filesystem, Qtree, Quota -- all these data models are unified from different vendors?! (There are exclusive parameters as well?)
> 5. From share side, what are we targetting?
> 6. Do we have any recommendations for the APIs when we considering a new type of (file) storage?! -- OR we are fitting these models to existing APIs?! -- This is a critical point and can impact the overall arch

1. In this PR we are modeling NAS/file storage resources for delfin
2. Ok
3. The 'security_mode' means operating system security mode (Windows, Unix) used by client to access shares/qtree
4. Like other resource in delfin we are collecting common/generic parameters. We are not handling vendor specific parameters
5. We are collecting CIFS (Windows) and NFS (Unix) models for Shares
6. We are creating new API for each resource (Eg. List). We are updating some APIs to make use of new APIs (Eg. Sync)

**Review by NajmudheenCT — APPROVED** [review]
LGTM

**Review by skdwriting — APPROVED** [review]
LGTM

**Pull request merged**